protocol/README.md

```diff
@@ -0,0 +1,405 @@
+
+### Request-Response
+
+Services may allow other services or clients to make a request against
+the service.
+
+Services must listen on the *request topic* `astoria/<slug>/request/+`
+where `<slug>` is the *slug* for the service. The wildcard at the end of
+the topic is for the *request slug*, a unique identifier for the
+request, e.g `kill_usercode`.
+
+A request can be made to a service by publishing a message to the
+appropriate *request topic*. The message may have the MQTT *response
+topic* set.
+
+The message content for a request must be a valid JSON object. The
+schema for the JSON object can vary depending on the request, and may be
+an empty object if no parameters are required.
+
+If the *response topic* is set on a request message, the service must
+respond by publishing a *response message* to the *response topic*. If
+no *response topic* is set, the service does not need to response, but
+should still attempt to perform the requested action.
+
+A *response topic* must be either of the format
+`astoria/<slug>/response/+` or be outside of the `astoria` topic
+namespace to avoid collisions.
+
+A *response message* must be a JSON object of the following format:
+
+``` json
+{
+    "protocol_version": "1.0.0",
+    "type": "response",
+    "service": "example",
+    "version": "1.2.3",
+    "success": true,
+    "message": "An optional user-facing message"
+}
+```
+
+-   `protocol_version` - protocol version for the service
+-   `type` - always `response` for response messages.
+-   `service` - *slug* for the service
+-   `version` - software version for the service
+-   `success`- boolean value, `true` if the request was successful,
+    `false` if not.
+-   `message` - optionally, a message for the user. This attribute may
+    be excluded from the message.
+
+### Streams
+
+A stream is a series of stream messages published to a given topic. A
+stream message can be published by clients or services.
+
+A stream message must only be acted on by clients. A service should be
+actioned using a request instead.
+
+A stream message is published to `astoria/stream/<stream_slug>`. It can
+be published by any component and multiple components can publish to the
+same stream simultaneously.
+
+Every stream must have a unique `stream_slug` that identifies the stream
+and its purpose, e.g `usercode_log`.
+
+A stream message must be a JSON object of the format:
+
+``` json
+{
+    "protocol_version": "1.0.0",
+    "type": "stream",
+    "stream": "example",
+    "sender": "example_sender",
+    "data": {"some_data": 1}
+}
+```
+
+-   `protocol_version` - protocol version
+-   `type` - always `stream` for stream messages.
+-   `stream` - *slug* for the stream
+-   `sender`- MQTT client name of the program that published the message
+-   `data` - The data for the event, varies by stream.
+
+## Defined Services
+
+Every service that operates as part of Astoria should be defined in this
+specification.
```

Comment:
Question: This implies that for any service to be within spec, it must be documented in the spec itself. That feels like a bad idea that won't scale. Instead, perhaps there should be either an ancillary document with each service defined, or the service should documents the details below itself, as it's only an implementation detail that these services are defined, rather than something encoded into the protocol itself.